qiskit_nature/operators/second_quantization/spin_op.py

```diff
@@ -12,117 +12,478 @@
+        print("This is same with: ", minus)
+
+        print("Dagger")
+        print(~(1j * z))
+
     """
-    Spin type operators. This class represents sums of `spin strings`, i.e. linear combinations of
-    SpinOperators with same spin and register length.
-    """
 
-    def __init__(self, operator_list):
-        # 1. Parse input
-        self._register_length = len(operator_list[0])
-        self.spin = operator_list[0].spin
-        for elem in operator_list:
-            assert isinstance(elem, SpinOperator)
-            assert len(elem) == self._register_length, \
-                'Cannot sum operators acting on registers of different length'
-            assert elem.spin == self.spin, \
-                'Cannot sum operators with different spins.'
+    _XYZ_DICT = {"X": 0, "Y": 1, "Z": 2}
+    _VALID_LABEL_PATTERN = re.compile(
+        r"^([IXYZ\+\-]_\d(\^\d)?\s)*[IXYZ\+\-]_\d(\^\d)?(?!\s)$|^[IXYZ\+\-]+$"
+    )
+    _SPARSE_LABEL_PATTERN = re.compile(r"^([IXYZ]_\d(\^\d)?\s)*[IXYZ]_\d(\^\d)?(?!\s)$")
+    _DENSE_LABEL_PATTERN = re.compile(r"^[IXYZ]+$")
+
+    def __init__(
+        self,
+        data: Union[
+            str,
+            List[Tuple[str, complex]],
+            Tuple[np.ndarray, np.ndarray],
+        ],
+        spin: Union[float, Fraction] = Fraction(1, 2),
+    ):
+        r"""
+        Args:
+            data: label string, list of labels and coefficients. See the label section in
+                  the documentation of :class:`SpinOp` for more details.
+            spin: positive half-integer (integer or half-odd-integer) that represents spin.
+
+        Raises:
+            ValueError: invalid data is given.
+            QiskitNatureError: invalid spin value
+        """
+        self._coeffs: np.ndarray
+        self._spin_array: np.ndarray
+        dtype = np.complex128  # TODO: configurable data type. mixin?
+
+        spin = Fraction(spin)
+        if spin.denominator not in (1, 2):
+            raise QiskitNatureError(
+                f"spin must be a positive half-integer (integer or half-odd-integer), not {spin}."
+            )
+        self._dim = int(2 * spin + 1)
+
+        if isinstance(data, tuple):
+            self._spin_array = np.array(data[0], dtype=np.uint8)
+            self._register_length = self._spin_array.shape[2]
+            self._coeffs = np.array(data[1], dtype=dtype)
 
-        # 2. Initialize the operator list of `self`
-        self._operator_list = operator_list
+        if isinstance(data, str):
+            data = [(data, 1)]
 
-        # 3. Set the operators particle type to 'spin S' with S the spin value (as float with 1
-        # decimal).
-        # SumOp.__init__(self, particle_type='spin {0:.1f}'.format(self.spin))
+        if isinstance(data, list):
+            invalid_labels = [label for label, _ in data if self._VALID_LABEL_PATTERN.match(label)]
+            if not invalid_labels:
+                raise ValueError(f"Invalid labels: {invalid_labels}")
```

Comment:
`invalid_labels` seems to be storing valid labels here. Then, if there is any valid label and some invalid ones, the list is not empty and not valid_labels resolves to False and no error is thrown. Seems like we have not enough unit tests with invalid labels that check this behaviour.